### PR TITLE
[docker]: Automate sonic-mgmt container creation

### PR DESCRIPTION
**Pull request by theasianpianist**
Signed-off-by: Lawrence Lee <lawlee@microsoft.com>

### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?
Ease the process of setting up a new sonic-mgmt Docker container by automating it
#### How did you do it?
* Add `setup-container.sh` script to create sonic-mgmt container and configure the user inside the container
* Update README with script instructions
#### How did you verify/test it?
Run the script as described in the README and verified on the created container that the UID and GID match between container and host, and that the user inside the container has passwordless sudo access.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,6 +7,20 @@ Tools for managing, configuring and monitoring SONiC
 # CII Best Practices
 [![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/3933/badge)](https://bestpractices.coreinfrastructure.org/projects/3933)
 
+# Docker Container Setup
+Use the `setup-container.sh` script to automatically create and configure your sonic-mgmt Docker container. This script should be run directly from the root of the sonic-mgmt repo. You should also run this script as the user that will be using the created container.
```
Comment (lguohan) [inline README.md:11]: how is this setup instruction related to the vssetup readme.md in the doc directory?

File: setup-container.sh [hunk 2]
```diff
@@ -0,0 +1,96 @@
+#! /bin/bash
+
+function show_help_and_exit() {
+    echo "Usage $0 [options]"
+    echo "Options with (*) are required"
+    echo "-h -?                 : get this help"
+    echo "-n <container name>   : (*) set the name of the Docker container"
+    echo "-i <image ID>         : (*) specify Docker image to use"
```
Comment (neethajohn) [inline setup-container.sh:8]: Can you put in an internal PR to also have a default image if user doesn't specify any since many users would just want to have a docker setup for running tests and would want to avoid the hassle of finding the right image to use?

Comment (daall) [reply setup-container.sh:8]: I think most people are using the public mgmt docker anyways so you could probably get away with pulling it from here: https://sonic-jenkins.westus2.cloudapp.azure.com/job/bldenv/job/docker-sonic-mgmt/lastSuccessfulBuild/artifact/sonic-buildimage/target/

Comment (theasianpianist) [reply setup-container.sh:8]: I'm not 100% sure how to approach this. As far as I can tell, there's no way to grab the image ID of the latest image from the public Jenkins, which means there's no way to check if that image already exists on the server before downloading it. If the image does exist already, Docker will automatically remove the name/tag from an existing image named "docker-sonic-mgmt" if it exists, which would result in possibly overwriting internally built images. This could also potentially result in many untagged images taking up disk space.

File: README.md (hunk 1 above)
Comment (theasianpianist) [reply README.md:11]: I don't believe they are really that related. The VSSetup document only applies to the virtual switch testbed, but the container created using this script is able to be used with physical devices. @daall please confirm/deny, I'm not entirely sure

Comment (theasianpianist) [reply README.md:11]: Nevermind, after talking with @daall I will update the VS Setup Readme to use this script so that the container setup is consistent.

File: setup-container.sh (hunk 2 above)
Comment (daall) [reply setup-container.sh:8]: That's a good point! We can revisit that later, I think the script is still helpful to have even without the auto-download step. :)

Comment (theasianpianist) [reply setup-container.sh:8]: @lguohan Having a docker registry where we can store the latest public image would make this feature pretty simple to implement (as suggested by @daall). Would it be possible to set one up and upload public sonic-mgmt images after they're built?

File: README.md (hunk 1 above)
Comment (shubav) [reply README.md:11]: I am little confused with the motivation of this PR. Can someone help me clear this confusion?

Do these changes add support for:
1. using docker VS instead of VM-based VS in the Vswitch test bed (or)
2. keeping the vm-based VS in the VS testbed but using docker VS instead of cEOS dockers for the other nodes

Apologies if this is covered elsewhere. I was unable to find it and understand the context. Thanks.

Comment (theasianpianist) [reply README.md:11]: Sorry for any confusion. This docker container is not meant to replace the VS testbed or any other testbed. The container created by this script is for the `docker-sonic-mgmt.gz` image, and is used for running tests on testbeds (both physical and virtual).

File: README.md [hunk 3]
```diff
@@ -7,6 +7,20 @@ Tools for managing, configuring and monitoring SONiC
...
+```
+Usage ./setup-container.sh [options]
+Options with (*) are required
+-h -?                 : get this help
+-n <container name>   : (*) set the name of the Docker container 
+-i <image ID>         : (*) specify Docker image to use
+-d <directory>        : specify directory inside container to bind mount to sonic-mgmt root (default "/var/src/")
+```
+
+After running the script, you should be able to enter the container using the `-u` option and your username.
```
Comment (daall) [inline README.md:22]: Can you add an example of the expected `docker exec -it` command here?

File: setup-container.sh [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            ;;
+        i )
+            IMAGE_ID=${OPTARG}
+            ;;
+        d )
+            LINK_DIR=${OPTARG}
+    esac
+done
+
+validate_parameters
```
Comment (daall) [inline setup-container.sh:104]: It may be a good idea to validate that the user is in the `docker` group by running `id docker` before starting the script.

File: setup-container.sh (hunk 2 above)
Comment (lguohan) [reply setup-container.sh:8]: yes, i believe we are already doing that.
